source/grammar/qasm3.g4

```diff
@@ -637,6 +641,15 @@ Newline : [\r\n]+ -> skip ;
 fragment Digit : [0-9] ;
 Integer : Digit+ ;
 
+fragment HexDigit : [0-9a-fA-F] ;
+fragment OctDigit : [0-7] ;
+fragment BinDigit : [01] ;
+fragment BinSequence : ( BinDigit [_]? )* BinDigit ;
+HexIntegerLiteral : ( '0x' | '0X' ) ( HexDigit [_]? )* HexDigit ;
+DecIntegerLiteral : ( Digit [_]? )* Digit ;
```

Comment:
The addition of `DecIntegerLiteral` makes some things a bit weird because it's not exactly the same as `Integer`.  All `Integer`s are valid `DecIntegerLiteral`s (but not vice-versa), and because `Integer` is defined first, ANTLR will always tokenise to `Integer` preferentially.  To avoid weird duplication in `expressionTerminator`, I think we might want to:
- make `Integer` a `fragment` rule
- remove `Integer` from `expressionTerminator`
- rewrite the `version` rule
- remove the `Integer` handling from the reference parser's `visitExpressionTerminator` (since the specific literals will now handle it)

The last point is currently tricky to do right (e.g. it currently allows `RealNumber` - `OPENQASM 3e+0;`!), so for now, I'd just set
```antlr
version: 'OPENQASM' DecIntegerLiteral ('.' DecIntegerLiteral)? SEMICOLON;
```
(which isn't ideal because it'd technically allow underscores in the version number).  Once #278 is in place, I can fix things in the reference implementation in a very neat way to tokenise the version number better (including potential alpha/beta/rc suffixes) without messing with the rest of the tokenisation.